ReadRows does not retry on all transient HTTP/2 failures

Example stack trace:

```java
com.google.api.gax.rpc.InternalException: io.grpc.StatusRuntimeException: INTERNAL: HTTP/2 error code: INTERNAL_ERROR
Received Rst Stream
    at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:67)
    at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
    at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
    at com.google.api.gax.grpc.ExceptionResponseObserver.onErrorImpl(ExceptionResponseObserver.java:82)
    at com.google.api.gax.rpc.StateCheckingResponseObserver.onError(StateCheckingResponseObserver.java:86)
    at com.google.api.gax.grpc.GrpcDirectStreamController$ResponseObserverAdapter.onClose(GrpcDirectStreamController.java:149)
    at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
    at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751)
    at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740)
    at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
    at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
    Suppressed: java.lang.RuntimeException: Asynchronous task failed
        at com.google.api.gax.rpc.ServerStreamIterator.hasNext(ServerStreamIterator.java:105)
        at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:42)
        at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
        at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439)
        at org.apache.spark.util.collection.ExternalSorter.insertAll(ExternalSorter.scala:191)
        at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:62)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:55)
        at org.apache.spark.scheduler.Task.run(Task.scala:123)
        at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
        ... 3 more
Caused by: io.grpc.StatusRuntimeException: INTERNAL: HTTP/2 error code: INTERNAL_ERROR
Received Rst Stream
    at io.grpc.Status.asRuntimeException(Status.java:533)
    ... 16 more
```

--- a/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/stub/readrows/ApiResultRetryAlgorithm.java
+++ b/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/stub/readrows/ApiResultRetryAlgorithm.java
@@ -29,14 +29,19 @@ public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<
   // Duration to sleep on if the error is DEADLINE_EXCEEDED.
   public static final Duration DEADLINE_SLEEP_DURATION = Duration.ofMillis(1);
 
+  private boolean isRetryableStatus(Status status) {
+    return status.getCode() == Status.Code.INTERNAL
+        && status.getDescription() != null
+        && (status.getDescription().contains("Received unexpected EOS on DATA frame from server")
+        || status.getDescription().contains("Received Rst Stream"));
+  }
+
   @Override
   public TimedAttemptSettings createNextAttempt(
       Throwable prevThrowable, ResponseT prevResponse, TimedAttemptSettings prevSettings) {
     if (prevThrowable != null) {
       Status status = Status.fromThrowable(prevThrowable);
-      if (status.getCode() == Status.Code.INTERNAL
-          && status.getDescription() != null
-          && status.getDescription().equals("Received unexpected EOS on DATA frame from server")) {
+      if (isRetryableStatus(status)) {
         return TimedAttemptSettings.newBuilder()
             .setGlobalSettings(prevSettings.getGlobalSettings())
             .setRetryDelay(prevSettings.getRetryDelay())
@@ -54,9 +59,7 @@ public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<
   public boolean shouldRetry(Throwable prevThrowable, ResponseT prevResponse) {
     if (prevThrowable != null) {
       Status status = Status.fromThrowable(prevThrowable);
-      if (status.getCode() == Status.Code.INTERNAL
-          && status.getDescription() != null
-          && status.getDescription().equals("Received unexpected EOS on DATA frame from server")) {
+      if (isRetryableStatus(status)) {
         return true;
       }
     }

--- a/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/stub/readrows/ApiResultRetryAlgorithm.java
+++ b/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/stub/readrows/ApiResultRetryAlgorithm.java
@@ -33,7 +33,7 @@ public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<
     return status.getCode() == Status.Code.INTERNAL
         && status.getDescription() != null
         && (status.getDescription().contains("Received unexpected EOS on DATA frame from server")
-        || status.getDescription().contains("Received Rst Stream"));
+            || status.getDescription().contains("Received Rst Stream"));
   }
 
   @Override

--- a/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta1/stub/readrows/ApiResultRetryAlgorithm.java
+++ b/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta1/stub/readrows/ApiResultRetryAlgorithm.java
@@ -29,14 +29,19 @@ public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<
   // Duration to sleep on if the error is DEADLINE_EXCEEDED.
   public static final Duration DEADLINE_SLEEP_DURATION = Duration.ofMillis(1);
 
+  private boolean isRetryableStatus(Status status) {
+    return status.getCode() == Status.Code.INTERNAL
+        && status.getDescription() != null
+        && (status.getDescription().contains("Received unexpected EOS on DATA frame from server")
+            || status.getDescription().contains("Received Rst Stream"));
+  }
+
   @Override
   public TimedAttemptSettings createNextAttempt(
       Throwable prevThrowable, ResponseT prevResponse, TimedAttemptSettings prevSettings) {
     if (prevThrowable != null) {
       Status status = Status.fromThrowable(prevThrowable);
-      if (status.getCode() == Status.Code.INTERNAL
-          && status.getDescription() != null
-          && status.getDescription().equals("Received unexpected EOS on DATA frame from server")) {
+      if (isRetryableStatus(status)) {
         return TimedAttemptSettings.newBuilder()
             .setGlobalSettings(prevSettings.getGlobalSettings())
             .setRetryDelay(prevSettings.getRetryDelay())
@@ -54,9 +59,7 @@ public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<
   public boolean shouldRetry(Throwable prevThrowable, ResponseT prevResponse) {
     if (prevThrowable != null) {
       Status status = Status.fromThrowable(prevThrowable);
-      if (status.getCode() == Status.Code.INTERNAL
-          && status.getDescription() != null
-          && status.getDescription().equals("Received unexpected EOS on DATA frame from server")) {
+      if (isRetryableStatus(status)) {
         return true;
       }
     }

--- a/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/stub/readrows/ApiResultRetryAlgorithm.java
+++ b/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/stub/readrows/ApiResultRetryAlgorithm.java
@@ -29,14 +29,19 @@ public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<
   // Duration to sleep on if the error is DEADLINE_EXCEEDED.
   public static final Duration DEADLINE_SLEEP_DURATION = Duration.ofMillis(1);
 
+  private boolean isRetryableStatus(Status status) {
+    return status.getCode() == Status.Code.INTERNAL
+        && status.getDescription() != null
+        && (status.getDescription().contains("Received unexpected EOS on DATA frame from server")
+        || status.getDescription().contains("Received Rst Stream"));
+  }
+
   @Override
   public TimedAttemptSettings createNextAttempt(
       Throwable prevThrowable, ResponseT prevResponse, TimedAttemptSettings prevSettings) {
     if (prevThrowable != null) {
       Status status = Status.fromThrowable(prevThrowable);
-      if (status.getCode() == Status.Code.INTERNAL
-          && status.getDescription() != null
-          && status.getDescription().equals("Received unexpected EOS on DATA frame from server")) {
+      if (isRetryableStatus(status)) {
         return TimedAttemptSettings.newBuilder()
             .setGlobalSettings(prevSettings.getGlobalSettings())
             .setRetryDelay(prevSettings.getRetryDelay())
@@ -54,9 +59,7 @@ public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<
   public boolean shouldRetry(Throwable prevThrowable, ResponseT prevResponse) {
     if (prevThrowable != null) {
       Status status = Status.fromThrowable(prevThrowable);
-      if (status.getCode() == Status.Code.INTERNAL
-          && status.getDescription() != null
-          && status.getDescription().equals("Received unexpected EOS on DATA frame from server")) {
+      if (isRetryableStatus(status)) {
         return true;
       }
     }

--- a/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/stub/readrows/ApiResultRetryAlgorithm.java
+++ b/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/stub/readrows/ApiResultRetryAlgorithm.java
@@ -33,7 +33,7 @@ public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<
     return status.getCode() == Status.Code.INTERNAL
         && status.getDescription() != null
         && (status.getDescription().contains("Received unexpected EOS on DATA frame from server")
-        || status.getDescription().contains("Received Rst Stream"));
+            || status.getDescription().contains("Received Rst Stream"));
   }
 
   @Override

--- a/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/BigQueryReadClientTest.java
+++ b/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/BigQueryReadClientTest.java
@@ -17,16 +17,20 @@ package com.google.cloud.bigquery.storage.v1;
 
 import com.google.api.gax.core.NoCredentialsProvider;
 import com.google.api.gax.grpc.GaxGrpcProperties;
+import com.google.api.gax.grpc.GrpcStatusCode;
 import com.google.api.gax.grpc.testing.LocalChannelProvider;
 import com.google.api.gax.grpc.testing.MockGrpcService;
 import com.google.api.gax.grpc.testing.MockServiceHelper;
 import com.google.api.gax.grpc.testing.MockStreamObserver;
 import com.google.api.gax.rpc.ApiClientHeaderProvider;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.InternalException;
 import com.google.api.gax.rpc.InvalidArgumentException;
 import com.google.api.gax.rpc.ServerStreamingCallable;
 import com.google.api.gax.rpc.StatusCode;
 import com.google.protobuf.AbstractMessage;
 import io.grpc.Status;
+import io.grpc.Status.Code;
 import io.grpc.StatusRuntimeException;
 import java.io.IOException;
 import java.util.Arrays;
@@ -165,10 +169,36 @@ public class BigQueryReadClientTest {
 
   @Test
   @SuppressWarnings("all")
-  public void readRowsRetryingExceptionTest() throws ExecutionException, InterruptedException {
-    StatusRuntimeException exception =
-        new StatusRuntimeException(
-            Status.INTERNAL.withDescription("Received unexpected EOS on DATA frame from server"));
+  public void readRowsRetryingEOSExceptionTest() throws ExecutionException, InterruptedException {
+    ApiException exception =
+        new InternalException(
+            new StatusRuntimeException(
+                Status.INTERNAL.withDescription(
+                    "Received unexpected EOS on DATA frame from server")),
+            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+    mockBigQueryRead.addException(exception);
+    long rowCount = 1340416618L;
+    ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();
+    mockBigQueryRead.addResponse(expectedResponse);
+    ReadRowsRequest request = ReadRowsRequest.newBuilder().build();
+
+    MockStreamObserver<ReadRowsResponse> responseObserver = new MockStreamObserver<>();
+
+    ServerStreamingCallable<ReadRowsRequest, ReadRowsResponse> callable = client.readRowsCallable();
+    callable.serverStreamingCall(request, responseObserver);
+    List<ReadRowsResponse> actualResponses = responseObserver.future().get();
+    Assert.assertEquals(1, actualResponses.size());
+  }
+
+  @Test
+  @SuppressWarnings("all")
+  public void readRowsRetryingHttp2StreamRstTest() throws ExecutionException, InterruptedException {
+    ApiException exception =
+        new InternalException(
+            new StatusRuntimeException(
+                Status.INTERNAL.withDescription(
+                    "HTTP/2 error code: INTERNAL_ERROR\nReceived Rst Stream")),
+            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
     mockBigQueryRead.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();

--- a/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/BigQueryReadClientTest.java
+++ b/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/BigQueryReadClientTest.java
@@ -175,7 +175,8 @@ public class BigQueryReadClientTest {
             new StatusRuntimeException(
                 Status.INTERNAL.withDescription(
                     "Received unexpected EOS on DATA frame from server")),
-            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+            GrpcStatusCode.of(Code.INTERNAL),
+            /* retryable = */ false);
     mockBigQueryRead.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();
@@ -198,7 +199,8 @@ public class BigQueryReadClientTest {
             new StatusRuntimeException(
                 Status.INTERNAL.withDescription(
                     "HTTP/2 error code: INTERNAL_ERROR\nReceived Rst Stream")),
-            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+            GrpcStatusCode.of(Code.INTERNAL),
+            /* retryable = */ false);
     mockBigQueryRead.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();

--- a/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta1/BigQueryStorageClientTest.java
+++ b/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta1/BigQueryStorageClientTest.java
@@ -305,7 +305,8 @@ public class BigQueryStorageClientTest {
             new StatusRuntimeException(
                 Status.INTERNAL.withDescription(
                     "Received unexpected EOS on DATA frame from server")),
-            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+            GrpcStatusCode.of(Code.INTERNAL),
+            /* retryable = */ false);
     mockBigQueryStorage.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();
@@ -328,7 +329,8 @@ public class BigQueryStorageClientTest {
             new StatusRuntimeException(
                 Status.INTERNAL.withDescription(
                     "HTTP/2 error code: INTERNAL_ERROR\nReceived Rst Stream")),
-            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+            GrpcStatusCode.of(Code.INTERNAL),
+            /* retryable = */ false);
     mockBigQueryStorage.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();

--- a/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta1/BigQueryStorageClientTest.java
+++ b/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta1/BigQueryStorageClientTest.java
@@ -17,11 +17,14 @@ package com.google.cloud.bigquery.storage.v1beta1;
 
 import com.google.api.gax.core.NoCredentialsProvider;
 import com.google.api.gax.grpc.GaxGrpcProperties;
+import com.google.api.gax.grpc.GrpcStatusCode;
 import com.google.api.gax.grpc.testing.LocalChannelProvider;
 import com.google.api.gax.grpc.testing.MockGrpcService;
 import com.google.api.gax.grpc.testing.MockServiceHelper;
 import com.google.api.gax.grpc.testing.MockStreamObserver;
 import com.google.api.gax.rpc.ApiClientHeaderProvider;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.InternalException;
 import com.google.api.gax.rpc.InvalidArgumentException;
 import com.google.api.gax.rpc.ServerStreamingCallable;
 import com.google.api.gax.rpc.StatusCode;
@@ -40,6 +43,7 @@ import com.google.cloud.bigquery.storage.v1beta1.TableReferenceProto.TableRefere
 import com.google.protobuf.AbstractMessage;
 import com.google.protobuf.Empty;
 import io.grpc.Status;
+import io.grpc.Status.Code;
 import io.grpc.StatusRuntimeException;
 import java.io.IOException;
 import java.util.Arrays;
@@ -295,10 +299,36 @@ public class BigQueryStorageClientTest {
 
   @Test
   @SuppressWarnings("all")
-  public void readRowsRetryingExceptionTest() throws ExecutionException, InterruptedException {
-    StatusRuntimeException exception =
-        new StatusRuntimeException(
-            Status.INTERNAL.withDescription("Received unexpected EOS on DATA frame from server"));
+  public void readRowsRetryingEOSExceptionTest() throws ExecutionException, InterruptedException {
+    ApiException exception =
+        new InternalException(
+            new StatusRuntimeException(
+                Status.INTERNAL.withDescription(
+                    "Received unexpected EOS on DATA frame from server")),
+            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+    mockBigQueryStorage.addException(exception);
+    long rowCount = 1340416618L;
+    ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();
+    mockBigQueryStorage.addResponse(expectedResponse);
+    ReadRowsRequest request = ReadRowsRequest.newBuilder().build();
+
+    MockStreamObserver<ReadRowsResponse> responseObserver = new MockStreamObserver<>();
+
+    ServerStreamingCallable<ReadRowsRequest, ReadRowsResponse> callable = client.readRowsCallable();
+    callable.serverStreamingCall(request, responseObserver);
+    List<ReadRowsResponse> actualResponses = responseObserver.future().get();
+    Assert.assertEquals(1, actualResponses.size());
+  }
+
+  @Test
+  @SuppressWarnings("all")
+  public void readRowsRetryingHttp2StreamRstTest() throws ExecutionException, InterruptedException {
+    ApiException exception =
+        new InternalException(
+            new StatusRuntimeException(
+                Status.INTERNAL.withDescription(
+                    "HTTP/2 error code: INTERNAL_ERROR\nReceived Rst Stream")),
+            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
     mockBigQueryStorage.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();

--- a/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta2/BigQueryReadClientTest.java
+++ b/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta2/BigQueryReadClientTest.java
@@ -175,7 +175,8 @@ public class BigQueryReadClientTest {
             new StatusRuntimeException(
                 Status.INTERNAL.withDescription(
                     "Received unexpected EOS on DATA frame from server")),
-            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+            GrpcStatusCode.of(Code.INTERNAL),
+            /* retryable = */ false);
     mockBigQueryRead.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();
@@ -198,7 +199,8 @@ public class BigQueryReadClientTest {
             new StatusRuntimeException(
                 Status.INTERNAL.withDescription(
                     "HTTP/2 error code: INTERNAL_ERROR\nReceived Rst Stream")),
-            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+            GrpcStatusCode.of(Code.INTERNAL),
+            /* retryable = */ false);
     mockBigQueryRead.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();

--- a/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta2/BigQueryReadClientTest.java
+++ b/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1beta2/BigQueryReadClientTest.java
@@ -17,16 +17,20 @@ package com.google.cloud.bigquery.storage.v1beta2;
 
 import com.google.api.gax.core.NoCredentialsProvider;
 import com.google.api.gax.grpc.GaxGrpcProperties;
+import com.google.api.gax.grpc.GrpcStatusCode;
 import com.google.api.gax.grpc.testing.LocalChannelProvider;
 import com.google.api.gax.grpc.testing.MockGrpcService;
 import com.google.api.gax.grpc.testing.MockServiceHelper;
 import com.google.api.gax.grpc.testing.MockStreamObserver;
 import com.google.api.gax.rpc.ApiClientHeaderProvider;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.InternalException;
 import com.google.api.gax.rpc.InvalidArgumentException;
 import com.google.api.gax.rpc.ServerStreamingCallable;
 import com.google.api.gax.rpc.StatusCode;
 import com.google.protobuf.AbstractMessage;
 import io.grpc.Status;
+import io.grpc.Status.Code;
 import io.grpc.StatusRuntimeException;
 import java.io.IOException;
 import java.util.Arrays;
@@ -165,10 +169,36 @@ public class BigQueryReadClientTest {
 
   @Test
   @SuppressWarnings("all")
-  public void readRowsRetryingExceptionTest() throws ExecutionException, InterruptedException {
-    StatusRuntimeException exception =
-        new StatusRuntimeException(
-            Status.INTERNAL.withDescription("Received unexpected EOS on DATA frame from server"));
+  public void readRowsRetryingEOSExceptionTest() throws ExecutionException, InterruptedException {
+    ApiException exception =
+        new InternalException(
+            new StatusRuntimeException(
+                Status.INTERNAL.withDescription(
+                    "Received unexpected EOS on DATA frame from server")),
+            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
+    mockBigQueryRead.addException(exception);
+    long rowCount = 1340416618L;
+    ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();
+    mockBigQueryRead.addResponse(expectedResponse);
+    ReadRowsRequest request = ReadRowsRequest.newBuilder().build();
+
+    MockStreamObserver<ReadRowsResponse> responseObserver = new MockStreamObserver<>();
+
+    ServerStreamingCallable<ReadRowsRequest, ReadRowsResponse> callable = client.readRowsCallable();
+    callable.serverStreamingCall(request, responseObserver);
+    List<ReadRowsResponse> actualResponses = responseObserver.future().get();
+    Assert.assertEquals(1, actualResponses.size());
+  }
+
+  @Test
+  @SuppressWarnings("all")
+  public void readRowsRetryingHttp2StreamRstTest() throws ExecutionException, InterruptedException {
+    ApiException exception =
+        new InternalException(
+            new StatusRuntimeException(
+                Status.INTERNAL.withDescription(
+                    "HTTP/2 error code: INTERNAL_ERROR\nReceived Rst Stream")),
+            GrpcStatusCode.of(Code.INTERNAL), /* retryable = */ false);
     mockBigQueryRead.addException(exception);
     long rowCount = 1340416618L;
     ReadRowsResponse expectedResponse = ReadRowsResponse.newBuilder().setRowCount(rowCount).build();

